dlib/image_loader/webp_loader.h

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2010  Davis E. King (davis@dlib.net), Nils Labugt
+// License: Boost Software License   See LICENSE.txt for the full license.
+#ifndef DLIB_WEBP_IMPORT
+#define DLIB_WEBP_IMPORT
+
+#include <vector>
+
+#include "webp_loader_abstract.h"
+#include "image_loader.h"
+#include "../pixel.h"
+#include "../dir_nav.h"
+#include "../test_for_odr_violations.h"
+
+namespace dlib
+{
+
+    class webp_loader : noncopyable
+    {
+    public:
+
+        webp_loader( const char* filename );
+        webp_loader( const std::string& filename );
+        webp_loader( const dlib::file& f );
+        webp_loader( const unsigned char* imgbuffer, size_t buffersize );
+
+        template<typename T>
+        void get_image( T& t_ ) const
+        {
+#ifndef DLIB_WEBP_SUPPORT
+            /* !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+                You are getting this error because you are trying to use the webp_loader
+                object but you haven't defined DLIB_WEBP_SUPPORT.  You must do so to use
+                this object.   You must also make sure you set your build environment
+                to link against the libwebp library.
+            !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!*/
+            COMPILE_TIME_ASSERT(sizeof( T ) == 0);
+#endif
+            image_view<T> t( t_ );
+            t.set_size( height_, width_ );
+            typedef typename image_traits<T>::pixel_type pixel_type;
+
+            unsigned char *output = reinterpret_cast<unsigned char*>( image_data( t ) );
+            const int stride = width_step( t );
+            const size_t output_size = stride * height_;
+
+            if ( pixel_traits<pixel_type>::rgb_alpha )
+            {
+                read_rgba( output, output_size, stride );
+                return;
+            }
+            if ( pixel_traits<pixel_type>::bgr_alpha )
+            {
+                read_bgra( output, output_size, stride );
+                return;
+            }
+            if ( pixel_traits<pixel_type>::rgb )
+            {
+                read_rgb( output, output_size, stride );
+                return;
+            }
+            if ( pixel_traits<pixel_type>::bgr_layout )
+            {
+                read_bgr( output, output_size, stride );
+                return;
+            }
```

Comment:
For `bgr_pixel` this code will do the wrong thing, since `pixel_traits<bgr_pixel>::rgb==true`.  Shouldn't it be this?

```suggestion
            if (pixel_traits<pixel_type>::rgb_alpha)
            {
                if (pixel_traits<pixel_type>::bgr_layout)
                    read_bgra(output, output_size, stride);
                else             
                    read_rgba(output, output_size, stride);   
            }
            if (pixel_traits<pixel_type>::rgb)
            {
                if (pixel_traits<pixel_type>::bgr_layout)
                    read_bgr(output, output_size, stride);
                else             
                    read_rgb(output, output_size, stride);                    
                return;
            }
```